### PR TITLE
Standalone Mode (beta)

### PR DESCRIPTION
**Pull request by pmarkowsky**
This PR creates a new Santa operating mode -- *standalone*
    
This adds a new operating mode to Santa called standalone mode.
    
When running in standalone mode TouchID can be used to approved binaries. If a
binary is properly signed a SigningID rule is generated otherwise a SHA256 rule
is generated. Note this lacks a GUI for browsing rules in the local rule db.

Example video of behavior attached (has the wrong logo):

[Standalone Santa Operation](https://www.youtube.com/watch?v=ja5mylENUro)

This PR is marked draft until the following things are added

1. [x] Comments are cleaned up
2. [ ] Documentation is updated (Mostly done -- sync updated, mode updated)
3. [ ] Tests are updated
4. [x] We decide if this is the right place to put the auth (move to reply block)
5. [x] We decide if Events need to have a new flag on them that informs the sync service that a binary was approved via Standalone mode. (Added standaloneApproval flag to events.
6. [ ] Rebase changes for the Swift UI

---
Aside I've been running with this mode for a month+.

### CODE REVIEW

File: Source/gui/SNTBinaryMessageWindowController.m [hunk 1]
```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
                                                showGroup:YES];
 }
 
+// When running in standalone mode, the user is prompted to approve the binary.
+- (void)approveBinaryForStandaloneMode {
+  LAContext *context;
```
Comment (tburgin) [inline Source/gui/SNTBinaryMessageWindowController.m:155]: nit: remove this and just declare and initialize on L159

File: Source/gui/SNTBinaryMessageWindowController.m [hunk 2]
```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
...
+  NSError *err;
+
+  dispatch_semaphore_t sema = dispatch_semaphore_create(0);
```
Comment (tburgin) [inline Source/gui/SNTBinaryMessageWindowController.m:158]: move this to after the potential -canEvaluatePolicy early return

File: Source/gui/SNTBinaryMessageWindowController.m [hunk 3]
```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
...
+  context = [[LAContext alloc] init];
+
+  if (![context canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&err]) {
```
Comment (tburgin) [inline Source/gui/SNTBinaryMessageWindowController.m:161]: I would do this check even before changing the UI in -loadWindow. Or better yet, when standalone mode is enabled, but bio auth is not available, we display an error in the block popup letting the user know.

File: Source/gui/SNTBinaryMessageWindowController.m [hunk 4]
```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
...
+                          addStandaloneRule:newRule
+                                      reply:^(NSError *error) {
+                                        if (error) {
+                                          LOGE(@"Failed to modify rules standalone: %s",
+                                               [error.localizedDescription UTF8String]);
+                                          LOGE(@"Failure reason: %@", error.localizedFailureReason);
+                                        }
+                                        dispatch_semaphore_signal(sema2);
+                                      }];
+                        dispatch_semaphore_wait(sema2, DISPATCH_TIME_FOREVER);
```
Comment (tburgin) [inline Source/gui/SNTBinaryMessageWindowController.m:213]: We should add some timeouts.

File: Source/gui/SNTBinaryMessageWindowController.m [hunk 5]
```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
...
+                      } else {
+                        LOGE(@"Authentication Error: %@", error);
+                      }
+                      dispatch_semaphore_signal(sema);
+                    }];
+  dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
```
Comment (tburgin) [inline Source/gui/SNTBinaryMessageWindowController.m:219]: Same here

File: Source/gui/SNTBinaryMessageWindowController.m [hunk 6]
```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
...
+                          LOGE(@"Connection invalidated");
+                          dispatch_semaphore_signal(sema2);
+                        };
+
+                        [daemonConn resume];
+                        [[daemonConn remoteObjectProxy]
+                          addStandaloneRule:newRule
+                                      reply:^(NSError *error) {
+                                        if (error) {
+                                          LOGE(@"Failed to modify rules standalone: %s",
```
Comment (tburgin) [inline Source/gui/SNTBinaryMessageWindowController.m:207]: Need some user feedback in the UI for this case.

File: Source/gui/SNTBinaryMessageWindowController.m [hunk 7]
```diff
@@ -135,13 +150,96 @@ - (IBAction)showCertInfo:(id)sender {
...
+
   NSURL *url = [SNTBlockMessage eventDetailURLForEvent:self.event customURL:self.customURL];
 
   [self closeWindow:sender];
   [[NSWorkspace sharedWorkspace] openURL:url];
 }
 
+- (IBAction)checkEventDetails:(id)sender {
+  NSString *eventCheckURL =
+    [NSString stringWithFormat:@"https://www.virustotal.com/gui/search/%@", self.event.fileSHA256];
```
Comment (tburgin) [inline Source/gui/SNTBinaryMessageWindowController.m:237]: I like this idea. Though we should make it configurable. Similar to what we do for event details.

I don't think we want to hard code virus total here.

File: Source/gui/SNTBinaryMessageWindowController.m (hunk 4 above)
Comment (pmarkowsky) [reply Source/gui/SNTBinaryMessageWindowController.m:213]: Good call, should this be configurable?

File: Source/gui/SNTBinaryMessageWindowController.h [hunk 8]
```diff
@@ -59,6 +59,12 @@
 ///
 @property(weak) IBOutlet NSButton *dismissEventButton;
 
+///
+///   Reference to the "Check Event" button in the XIB. Used to help the user
+///   check the binary.
+///
+@property(weak) IBOutlet NSButton *checkEventButton;
+
```
Comment (pmarkowsky) [inline Source/gui/SNTBinaryMessageWindowController.h:67]: This was an accidental squash error, will replace this and move it to another PR.

File: Source/gui/SNTBinaryMessageWindowController.m (hunk 3 above)
Comment (pmarkowsky) [reply Source/gui/SNTBinaryMessageWindowController.m:159]: > Or better yet, when standalone mode is enabled, but bio auth is not available, we display an error in the block popup letting the user know

This seems like a good idea. Should we also disable the `Approve` button so users don't get frustrated?

File: Source/common/SNTXPCUnprivilegedControlInterface.h [hunk 9]
```diff
@@ -51,6 +51,11 @@ struct RuleCounts {
 - (void)databaseEventCount:(void (^)(int64_t count))reply;
 - (void)staticRuleCount:(void (^)(int64_t count))reply;
 
+///
+/// Standalone rules
+///
+- (void)addStandaloneRule:(SNTRule *)rule reply:(void (^)(NSError *))reply;
```
Comment (tburgin) [inline Source/common/SNTXPCUnprivilegedControlInterface.h:57]: I don’t think we want to expose adding arbitrary rules over the unprivileged interface. Instead, we could have the daemon drive the flow, sending a message to the GUI for auth.

Comment (pmarkowsky) [reply Source/common/SNTXPCUnprivilegedControlInterface.h:57]: Admittedly I went back and forth on this. What were you thinking the flow would look like?

The main reason I added this was to keep the flow simpler then the santad flows I sketched out. Felt like I had to keep state in the daemon which felt wrong. 

If you enable standalone mode then our flow looks like this.

```mermaid
sequenceDiagram
santad ->> gui: We blocked a binary pop up message and display approve button
gui ->> gui: User clicks approve
gui ->> gui: Get auth from user
gui -->> santad: addStandaloneRule
```

If the user clicks ignore then nothing happens.

File: Source/gui/SNTBinaryMessageWindowController.m (hunk 3 above)
Comment (tburgin) [reply Source/gui/SNTBinaryMessageWindowController.m:156]: Yeah, if the desired auth is not available we shouldn't offer standalone buttons.

File: Source/gui/SNTBinaryMessageWindowController.m (hunk 4 above)
Comment (tburgin) [reply Source/gui/SNTBinaryMessageWindowController.m:210]: Most other XPC calls we use a 5 sec timeout. That should work fine here too.

File: Source/common/SNTXPCUnprivilegedControlInterface.h (hunk 9 above)
Comment (tburgin) [reply Source/common/SNTXPCUnprivilegedControlInterface.h:57]: > The main reason I added this was to keep the flow simpler then the santad flows I sketched out. Felt like I had to keep state in the daemon which felt wrong.

Having the GUI push in rules is simpler, but I don't think we want to expose adding rules from an unprivileged interface. With that as a requirement, we will need to store state with the daemon.

sequenceDiagram
santad ->> santad: blocks an executable - checks if standalone mode and desired auth is available
santad ->> rules.db: adds a pending rule
santad ->> gui: prompt user and display approve button
gui ->> gui: User clicks approve
gui ->> gui: Get auth from user
gui ->> santad: activateStandaloneRule:
santad ->> rules.db: converts pending rule to active

The idea of a "pending rule" is new and we would need to think though the implications.

The other thing we need to consider is the key used with activateStandaloneRule. The obvious choice is to just use the blockable event. This is probably fine. However, it does not strongly guarantee that a user approved the event. We could consider using a hard to guess nonce, or figure out if LocalAuthentication can mint an attestation that we verify in the activateStandaloneRule: implementation.

Comment (tburgin) [reply Source/common/SNTXPCUnprivilegedControlInterface.h:57]: Chatted offline - using a reply block would eliminate the need for exposing an additional interface in the daemon. This is a great idea!

File: Source/gui/SNTBinaryMessageWindowController.m (hunk 4 above)
Comment (pmarkowsky) [reply Source/gui/SNTBinaryMessageWindowController.m]: I removed the semaphore since it's unnecessary now.

File: Source/gui/SNTBinaryMessageWindowController.m (hunk 3 above)
Comment (pmarkowsky) [reply Source/gui/SNTBinaryMessageWindowController.m]: I made the button disappear. But haven't figured out how to update the UI to display to the user that things are not operating as intended.

**Comment by pmarkowsky** [comment]
Going to close this PR and start a new one based on the new UI code.
